File: ex_plataforma_coleta_de_dados/PCD.cs
using System;
using System.Collections.Generic;

namespace SistemaMonitoramentoAmazonia
{
    public class PCD
    {
        private List<IObservadorPCD> _universidades = new List<IObservadorPCD>();
        private double _temperatura;
        private double _ph;
        private double _umidade;

        public PCD(string nome) => Nome = nome;

        public void RegistrarInteresse(IObservadorPDC uni)
        {
            if (!_universidades.Contains(uni))
            {
                _universidades.Add(uni);
            }
        }

        public void setMedicoes(double temp, double ph, double umidade)
        {
            _temperatura = temp;
            _ph = ph;
            _umidade = umidade;
            Notificar();
        }

        private void Notificar()
        {
            foreach (var uni in _universidades)
            {
                uni.Atualizar(Nome, _temperatura, _ph, _umidade);
            }
        }

    }
}

File: ex_PCD_IOC/PCD.py
from IObservadorPCD import IObservadorPCD
class PCD:
    def __init__(self, nome):
        self.nome = nome
        # IoC: A lista começa vazia. As dependências serão "injetadas" depois, externamente.
        self._observadores = []

    def registrar_interesse(self, observador: IObservadorPCD):
        """
        INJEÇÃO DE DEPENDÊNCIA:
        Em vez da PCD criar a universidade, nós 'injetamos' a universidade pronta aqui dentro.
        """
        if observador not in self._observadores:
            self._observadores.append(observador)

    #assim que os dados mudam, o modelo dispara o evento para notificar
    def set_medicoes(self, temp, ph, umidade):
        print(f"\n[Sinal de Sensor] {self.nome} captou novos dados.")
        self._notificar(temp, ph, umidade)

    def _notificar(self, temp, ph, umidade):
        """
        HOLLYWOOD PRINCIPLE: 'Don't call us, we'll call you'.
        Os observadores não perguntam os dados. A PCD 'liga de volta' 
        para eles quando algo muda. O modelo percorre os observadoes e chala o callback 'atualizar'
        """
        for obs in self._observadores:
            obs.atualizar(self.nome, temp, ph, umidade)